lib/std/os.zig

```diff
@@ -4152,6 +4156,71 @@ pub fn getFdPath(fd: fd_t, out_buffer: *[MAX_PATH_BYTES]u8) RealPathError![]u8 {
+            // or similar, there's not a good way to get this in FreeBSD.
+            // Instead, list all open files for the process and find the FD we're looking for.
+            switch (errno(system.sysctlnametomib("kern.proc.filedesc", &mib[0], &miblen))) {
+                0 => {},
+                EFAULT => unreachable, // stack is always a valid address
+                EINVAL => unreachable, // known good name
+                ENOMEM => unreachable, // known size, on the stack
+                EPERM => unreachable, // not setting new value
+                ENOENT => return error.NotSupported,
+                else => |err| return unexpectedErrno(err),
+            }
+
+            // If the buffer is too small, sysctl silently gives partial result,
+            // which may not contain our target FD.
+            // Get the approximate required size by using null for oldp/buf.
+            // Use that to change the error produced if we can't find the FD.
+            // TODO: Do this better with an allocator, alloca or stupidly large array?
+            var buflen: usize = 0;
+            sysctl(&mib, null, &buflen, null, 0) catch |err| switch (err) {
+                error.PermissionDenied => unreachable, // not setting value
+                error.NameTooLong => unreachable, // known good mib from sysctlnametomib
+                error.UnknownName => unreachable, // known good mib from sysctlnametomib
+                error.Unexpected => return error.Unexpected, // EINVAL/EISDIR/ENOTDIR should be unreachable
+                error.SystemResources => return error.SystemResources,
+            };
+
+            // 20 is completely arbitrary, should work to around 40 open files?
+            var buf: [MAX_PATH_BYTES * 20]u8 = undefined;
+
+            // If the buffer size required is larger than above, try with the partial result.
+            // If unsuccessful, signal it's a resource constraint rather than not finding the FD.
+            const result_error: RealPathError = if (buflen > buf.len) error.SystemResources else error.FileNotFound;
+
+            buflen = buf.len;
+
+            sysctl(&mib, &buf, &buflen, null, 0) catch |err| switch (err) {
+                error.PermissionDenied => unreachable, // not setting value
+                error.NameTooLong => unreachable, // known good mib from sysctlnametomib
+                error.UnknownName => unreachable, // known good mib from sysctlnametomib
+                error.Unexpected => return error.Unexpected, // EINVAL/EISDIR/ENOTDIR should be unreachable
+                error.SystemResources => return error.SystemResources,
+            };
+
+            var offset: usize = 0;
+            while (offset < buflen) {
+                const kf = @ptrCast(*system.kinfo_file, @alignCast(8, &buf[offset]));
+                if (kf.kf_fd == fd) {
+                    const end = mem.indexOfScalar(u8, &kf.path, 0) orelse MAX_PATH_BYTES - 1;
```

Comment:
Do I read this right that if the path is actually longer than `MAX_PATH_BYTES`, we'll truncate it? Wouldn't `error.NameTooLong` not be a more appropriate response here?

Reply:
`MAX_PATH_BYTES` is the actual kernel's max path length including the terminating null byte, so I don't expect this to ever happen. Probably segfault territory before we even get to return an error? But that might be clearer to readers so I'll change that, thanks.